python/smqtk/algorithms/saliency/rise.py

```diff
@@ -0,0 +1,117 @@
+
+        masks = np.empty((N, *input_size))
+
+        for i in tqdm(range(N), desc='Generating masks'):
+            # Random shifts
+            x = np.random.randint(0, cell_size[0])
+            y = np.random.randint(0, cell_size[1])
+            # Linear upsampling and random cropping
+            masks[i, :, :] = resize(grid[i], up_size, order=1, mode='reflect',
+                                    anti_aliasing=False)[x:x + input_size[0], y:y + input_size[1]]
+        self.masks = masks.reshape(-1, *input_size, 1)
+    
+    def augment(self, image_mat):
+        """
+        :param numpy.ndarray image_mat:
+            Image matrix to be augmented.
+
+        :return: A numpy arrays of augmented image matrices as well as masks
+            that indicate the regions in the augmented images that are
+            unmodified with respect to the input image (preserved regions).
+
+            Returned augmented images should be in the dimension format
+            [index, height, width [,channel]] with the the same data type as
+            the input image matrix.
+
+            Returned masks should be in the dimension format
+            [index, height, width] with the boolean data type.
+        :rtype: (numpy.ndarray, numpy.ndarray)
+        """
+        # If image is grayscale
+        if len(image_mat.shape) == 2:
+            image_mat = np.expand_dims(image_mat, 2).repeat(3, axis=2)
+        return self.masks * image_mat, self.masks
+            
+        
+class RISEGenerator (ImageSaliencyMapGenerator):
+    def generate(self, base_image, augmenter, descriptor_generator,
+                 blackbox):
+        """
+        Generate an image saliency heat-map matrix given a blackbox's behavior
+        over the descriptions of an augmented base image.
+
+        :param numpy.ndarray image_mat:
+            Numpy image matrix of the format [height, width [,channel]] that is
+            to be augmented.
+
+        :param ImageSaliencyAugmenter augmenter:
+            Augmentation algorithm following
+            the :py:class:`ImageSaliencyAugmenter` interface.
+
+        :param smqtk.algorithms.DescriptorGenerator descriptor_generator:
+            A descriptor generation algorithm following
+            the :py:class:`smqtk.algorithms.DescriptorGenerator` interface.
+
+        :param SaliencyBlackbox blackbox:
+            Blackbox algorithm implementation following
+            the :py:class:`SaliencyBlackbox` interface.
+
+        :return: A :py:class:`numpy.ndarray` matrix of the same [height, width]
+            shape as the input image matrix but of floating-point type within
+            the range of [0,1], where areas of higher value represent more
+            salient regions according to the given blackbox algorithm.
+        :rtype: numpy.ndarray[float]
+        """
+        # augmenter = RISEAugmenter(1000, 10, 0.5, (224, 224))
+        masked_images, masks = augmenter.augment(base_image)
```

Comment:
Here we assume base_image height and width is equal to (224,224), which is not always true. Either resize it to (224,224) before it gets passed to the augmenter.augment() and resize it back to original height and width after getting the saliency map or pass image size as parameter every time we receive a new base_image. We chose to resize the images to (224,224) for other methods (for the time being). Please refer to the class "Logit_ImageSaliencyMapGenerator" on saliency-implementations branch for a closer look.